src/order/heyting_algebra.lean

```diff
@@ -0,0 +1,678 @@
+{ ..‹generalized_heyting_algebra α› }
+
+@[priority 100] -- See note [lower instance priority]
+instance generalized_coheyting_algebra.to_order_bot [generalized_coheyting_algebra α] :
+  order_bot α :=
+{ ..‹generalized_coheyting_algebra α› }
+
+@[priority 100] -- See note [lower instance priority]
+instance heyting_algebra.to_bounded_order [heyting_algebra α] : bounded_order α :=
+{ ..‹heyting_algebra α› }
+
+@[priority 100] -- See note [lower instance priority]
+instance coheyting_algebra.to_bounded_order [coheyting_algebra α] : bounded_order α :=
+{ ..‹coheyting_algebra α› }
+
+@[priority 100] -- See note [lower instance priority]
+instance biheyting_algebra.to_coheyting_algebra [biheyting_algebra α] : coheyting_algebra α :=
+{ ..‹biheyting_algebra α› }
+
+/-- Construct a Heyting algebra from the Heyting implication alone. -/
+@[reducible] -- See note [reducible non instances]
+def heyting_algebra.of_himp [distrib_lattice α] [bounded_order α] (himp : α → α → α)
+  (le_himp_iff : ∀ a b c, a ≤ himp b c ↔ a ⊓ b ≤ c) : heyting_algebra α :=
+{ himp := himp,
+  compl := λ a, himp a ⊥,
+  le_himp_iff := le_himp_iff,
+  himp_bot := λ a, rfl,
+  ..‹distrib_lattice α›, ..‹bounded_order α› }
+
+/-- Construct a Heyting algebra from the Heyting implication alone. -/
+@[reducible] -- See note [reducible non instances]
+def heyting_algebra.of_compl [distrib_lattice α] [bounded_order α] (compl : α → α)
+  (le_himp_iff : ∀ a b c, a ≤ compl b ⊔ c ↔ a ⊓ b ≤ c) : heyting_algebra α :=
+{ himp := λ a, (⊔) (compl a),
+  compl := compl,
+  le_himp_iff := le_himp_iff,
+  himp_bot := λ a, sup_bot_eq,
+  ..‹distrib_lattice α›, ..‹bounded_order α› }
+
+/-- Construct a co-Heyting algebra from the difference alone. -/
+@[reducible] -- See note [reducible non instances]
+def coheyting_algebra.of_sdiff [distrib_lattice α] [bounded_order α] (sdiff : α → α → α)
+  (sdiff_le_iff : ∀ a b c, sdiff a b ≤ c ↔ a ≤ b ⊔ c) : coheyting_algebra α :=
+{ sdiff := sdiff,
+  hnot := λ a, sdiff ⊤ a,
+  sdiff_le_iff := sdiff_le_iff,
+  top_sdiff := λ a, rfl,
+  ..‹distrib_lattice α›, ..‹bounded_order α› }
+
+/-- Construct a co-Heyting algebra from the Heyting negation alone. -/
+@[reducible] -- See note [reducible non instances]
+def coheyting_algebra.of_hnot [distrib_lattice α] [bounded_order α] (hnot : α → α)
+  (sdiff_le_iff : ∀ a b c, a ⊓ hnot b ≤ c ↔ a ≤ b ⊔ c) : coheyting_algebra α :=
+{ sdiff := λ a b, (a ⊓ hnot b),
+  hnot := hnot,
+  sdiff_le_iff := sdiff_le_iff,
+  top_sdiff := λ a, top_inf_eq,
+  ..‹distrib_lattice α›, ..‹bounded_order α› }
+
+section generalized_heyting_algebra
+variables [generalized_heyting_algebra α] {a b c d : α}
+
+-- `p → q → r ↔ p ∧ q → r`
+@[simp] lemma le_himp_iff : a ≤ b ⇨ c ↔ a ⊓ b ≤ c := generalized_heyting_algebra.le_himp_iff _ _ _
+
+-- `p → q → r ↔ q → p → r`
+lemma le_himp_comm : a ≤ b ⇨ c ↔ b ≤ a ⇨ c := by rw [le_himp_iff, le_himp_iff, inf_comm]
+
+-- `p → q → p`
+lemma le_himp : a ≤ b ⇨ a := le_himp_iff.2 inf_le_left
```

Comment:
What is your motivation for leaving most, but not all, variables implicit in this set of new lemmas?

Reply:
I came to the conclusion that simple lemmas stating an inequality should only have implicit arguments because empirically they are used in `le_trans` or such, where there's enough context for Lean to infer everything. This is different from equality lemmas, because those you can rewrite by, and rewriting gives a lot less context.

I seem to have left a few explicit arguments standing in the `≤` lemmas however. Let me fix this quick.